android/runtime/v8/src/java/org/appcelerator/kroll/runtime/v8/V8Promise.java

```diff
@@ -0,0 +1,90 @@
+package org.appcelerator.kroll.runtime.v8;
+
+import android.os.Message;
+
+import org.appcelerator.kroll.KrollPromise;
+import org.appcelerator.kroll.KrollRuntime;
+import org.appcelerator.kroll.common.AsyncResult;
+import org.appcelerator.kroll.common.TiMessenger;
+
+/**
+ * This class wraps an underlying v8::Promise::Resolver which holds
+ * the Promise but allows us to call resolve/reject on it.
+ */
+public abstract class V8Promise<V extends Object> extends V8Object implements KrollPromise<V>
+{
+	private static final String TAG = "V8Promise";
+
+	protected static final int MSG_RESOLVE = V8Object.MSG_LAST_ID + 100;
+	protected static final int MSG_REJECT = V8Object.MSG_LAST_ID + 101;
+	protected static final int MSG_LAST_ID = MSG_REJECT;
+
+	public V8Promise()
+	{
+		super(nativeCreate());
+		TiMessenger.postOnRuntime(this);
+	}
+
+	@Override
+	public void resolve(V value)
+	{
+		if (KrollRuntime.getInstance().isRuntimeThread()) {
+			nativeResolve(getPointer(), value);
+		} else {
+			TiMessenger.sendBlockingRuntimeMessage(handler.obtainMessage(MSG_RESOLVE), value);
+		}
+	}
+
+	@Override
+	public void reject(Object value)
+	{
+		if (KrollRuntime.getInstance().isRuntimeThread()) {
+			nativeReject(getPointer(), value);
+		} else {
+			TiMessenger.sendBlockingRuntimeMessage(handler.obtainMessage(MSG_REJECT), value);
+		}
+	}
+
+	@Override
+	public boolean handleMessage(Message message)
+	{
+		switch (message.what) {
+			case MSG_RESOLVE: {
+				AsyncResult asyncResult = ((AsyncResult) message.obj);
+				Object value = asyncResult.getArg();
+				nativeResolve(getPointer(), value);
+				asyncResult.setResult(null);
+				return true;
+			}
+			case MSG_REJECT: {
+				AsyncResult asyncResult = ((AsyncResult) message.obj);
+				Object value = asyncResult.getArg();
+				nativeReject(getPointer(), value);
+				asyncResult.setResult(null);
+				return true;
+			}
+		}
+
+		return super.handleMessage(message);
+	}
+
+	@Override
+	public void doRelease()
+	{
+		long resolverPointer = getPointer();
```

Comment:
I don't think this is necessary as [super.doRelease()](https://github.com/appcelerator/titanium_mobile/blob/master/android/runtime/v8/src/java/org/appcelerator/kroll/runtime/v8/V8Object.java#L79) will release the pointer.

###### NOTE: Can be removed from `V8Function` too.

Reply:
I think the issue here was that the `nativeRelease` I needed to call was the custom one for V8Promise because we keep the underlying Resolvers in a separate map. `V8Object`'s nativeRelease is defined as static and private, so I couldn't override it in the subclass.

So, I suppose I could eliminate this and just override the `nativeRelease` if we make it non-static and protected in `V8Object`. Same goes for `V8Function`.